finality-aleph/src/sync/handler/request_handler.rs

```diff
@@ -0,0 +1,423 @@
+                _ => state,
+            };
+
+            if *to == header.id() {
+                state = State::Finished;
+                break;
+            }
+
+            if to.number() >= header.id().number() {
+                return Err(RequestHandlerError::RootMismatch);
+            }
+
+            head_chunk = match state {
+                State::EverythingButHeader => {
+                    let (j, b) = self.get_justification_and_block(header.id())?;
+                    match j {
+                        Some(j) => HeadOfChunk::Justification(j, b, None),
+                        None => {
+                            pre_chunk.add_block(b);
+                            HeadOfChunk::Header(self.parent_header(&header)?)
+                        }
+                    }
+                }
+                State::Everything => {
+                    let (j, b, h) = self.get_justification_block_and_header(header.id())?;
+                    match j {
+                        Some(j) => HeadOfChunk::Justification(j, b, h),
+                        None => {
+                            pre_chunk.add_block(b);
+                            pre_chunk.add_header(h);
+                            HeadOfChunk::Header(self.parent_header(&header)?)
+                        }
+                    }
+                }
+                State::OnlyJustification => {
+                    let j = self.get_justification(header.id())?;
+                    match j {
+                        Some(j) => HeadOfChunk::Justification(j, None, None),
+                        None => HeadOfChunk::Header(self.parent_header(&header)?),
+                    }
+                }
+                State::Finished => return Err(RequestHandlerError::BadState),
+            };
+
+            state = match branch_knowledge {
+                BranchKnowledge::LowestId(id) if *id == header.id() => State::Everything,
+                _ => state,
+            };
+
+            if let HeadOfChunk::Justification(_, _, _) = &head_chunk {
+                break;
+            }
+        }
+
+        Ok((state, head_chunk, pre_chunk.into_chunks()))
+    }
+
+    fn chunks(
+        self,
+        head: HeadOfChunk<B, J>,
+        branch_knowledge: BranchKnowledge<J>,
+        to: BlockIdFor<J>,
+    ) -> HandlerResult<Vec<Chunk<B, J>>, CS::Error> {
+        let mut chunks = vec![];
+
+        let mut head = head;
+        let mut state = State::EverythingButHeader;
+
+        while state != State::Finished {
+            let (new_state, new_head, chunks_from_step) =
+                self.step(state, head, &to, &branch_knowledge)?;
+
+            chunks.extend(chunks_from_step);
+            state = new_state;
+            head = new_head;
+        }
+
+        Ok(chunks)
+    }
+
+    pub fn response(self, request: Request<J>) -> HandlerResult<Response<B, J>, CS::Error> {
+        if !request.is_valid() {
+            return Err(RequestHandlerError::BadRequest);
```

Comment:
we don't need this check, right? maybe remove it completely?

Reply:
done